ext/git.c

```diff
@@ -0,0 +1,218 @@
+    DDTRACE_G(git_metadata).repository_url = zend_string_copy(repository_url);
+}
+
+static bool add_git_info(zval* meta, zend_string* commit_sha, zend_string* repository_url, bool is_root_span, bool cache) {
+    if (commit_sha && repository_url && ZSTR_LEN(commit_sha) > 0 && ZSTR_LEN(repository_url) > 0) {
+        removeCredentials(repository_url);
+
+        if (is_root_span) {
+            add_assoc_str(meta, "_dd.git.commit.sha", zend_string_copy(commit_sha));
+            add_assoc_str(meta, "_dd.git.repository_url", zend_string_copy(repository_url));
+        } else {
+            add_assoc_str(meta, "git.commit.sha", zend_string_copy(commit_sha));
+            add_assoc_str(meta, "git.repository_url", zend_string_copy(repository_url));
+        }
+
+        if (cache) {
+            cache_git_metadata(commit_sha, repository_url);
+        }
+
+        return true;
+    }
+
+    return false;
+}
+
+bool inject_from_env(zval* meta, bool is_root_span) {
+    return add_git_info(meta, get_DD_GIT_COMMIT_SHA(), get_DD_GIT_REPOSITORY_URL(), is_root_span, true);
+}
+
+bool inject_from_global_tags(zval* meta, bool is_root_span) {
+    zend_array* global_tags = get_DD_TAGS();
+    bool success = false;
+
+    if (global_tags) {
+        zval* git_commit_sha = zend_hash_str_find(global_tags, ZEND_STRL("git.commit.sha"));
+        zval* git_repository_url = zend_hash_str_find(global_tags, ZEND_STRL("git.repository_url"));
+
+        if (git_commit_sha && git_repository_url && Z_TYPE_P(git_commit_sha) == IS_STRING &&
+            Z_TYPE_P(git_repository_url) == IS_STRING) {
+            success = add_git_info(meta, Z_STR_P(git_commit_sha), Z_STR_P(git_repository_url), is_root_span, true);
+        }
+    }
+
+    if (success && is_root_span) {
+        zend_hash_str_del(Z_ARR_P(meta), ZEND_STRL("git.commit.sha"));
+        zend_hash_str_del(Z_ARR_P(meta), ZEND_STRL("git.repository_url"));
+    }
+
+    return success;
+}
+
+bool inject_from_git_files(zval* meta, bool is_root_span) {
+    char cwd[PATH_MAX];
+    if (!getcwd(cwd, sizeof(cwd))) {
+        return false;
+    }
+
+    char git_dir[PATH_MAX];
+    snprintf(git_dir, sizeof(git_dir), "%s/.git", cwd);
+
+    zend_string* git_commit_sha = get_commit_sha(git_dir);
+    zend_string* git_repository_url = get_repository_url(git_dir);
+
+    if (!git_commit_sha || !git_repository_url) {
+        if (git_commit_sha) zend_string_release(git_commit_sha);
+        if (git_repository_url) zend_string_release(git_repository_url);
+        return false;
+    }
+
+    bool result = add_git_info(meta, git_commit_sha, git_repository_url, is_root_span, true);
+
+    zend_string_release(git_commit_sha);
+    zend_string_release(git_repository_url);
+
+    return result;
+}
+
+void ddtrace_inject_git_metadata(zval* meta, bool is_root_span) {
+    ddtrace_git_metadata git_metadata = DDTRACE_G(git_metadata);
+    if (git_metadata.called_once) {
```

Comment:
Do I see read it correctly, that this is only ever called once for the whole thread?
I suppose having multiple virtual hosts with multiple git repositories on an apache server is not that uncommon. So I suppose we'll need some HashMap based caching here?

Reply:
> By the way, called_once seems never set to true

Oops, I think I deleted the line 😬